### PR TITLE
Feat/montekki/implement protocol connection

### PR DESCRIPTION
**Pull request by montekki**
this is currently a WIP

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -20,35 +25,64 @@ type Server struct {
...
-		zap.S().Error("failed to send handshake: ", err)
-		return
-	}
-	_, err = handshake.ReadFrom(conn)
-	if err != nil {
-		zap.S().Error("failed to read handshake: ", err)
-		return
+func dialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	var major, minor, patch uint32
+	minor = 5
```
Comment (kazhuravlev) [inline pkg/server/server.go]: this is a constant. don`t you think so?

File: pkg/server/server.go [hunk 2]
```diff
@@ -20,35 +25,64 @@ type Server struct {
...
+
+	for i := 0; i < 20; i++ {
+		patch = uint32(i)
+		time.Sleep(20 * time.Second)
+		conn, err := dialer.DialContext(ctx, network, addr)
+		if err != nil {
+			continue
+		}
+
+		fmt.Printf("Trying to connect with version %v %v %v", major, minor, patch)
```
Comment (kazhuravlev) [inline pkg/server/server.go]: will be better if you using a logger instead fmt

File: pkg/server/server.go [hunk 3]
```diff
@@ -20,35 +25,64 @@ type Server struct {
...
-		return
+func dialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	var major, minor, patch uint32
+	minor = 5
+
+	dialer := net.Dialer{}
+
+	for i := 0; i < 20; i++ {
+		patch = uint32(i)
+		time.Sleep(20 * time.Second)
```
Comment (kazhuravlev) [inline pkg/server/server.go]:
```suggestion
		ticker := time.NewTimer(20*time.Second)
		
		select {
		case <- ticker.C:
		case <-ctx.Done():
		    return nil, ctx.Err()
		}
```

File: pkg/server/server.go [hunk 4]
```diff
@@ -20,35 +25,64 @@ type Server struct {
...
 	}
+	return nil, errors.New("TODO")
+}
 
-	var b []byte
-	b, e := json.Marshal(handshake)
-	if e != nil {
+func handleClient(peer string) {
+	customTransport := p2p.Transport{DialContext: dialContext}
+	conn, _ := p2p.NewConn(
```
Comment (kazhuravlev) [inline pkg/server/server.go]: handle `err`

File: pkg/server/server.go [hunk 5]
```diff
@@ -20,35 +25,64 @@ type Server struct {
...
+		p2p.WithVersion(proto.Version{Major: 0, Minor: 5, Patch: 14}),
```
Comment (kazhuravlev) [inline pkg/server/server.go]: repeated constant

File: pkg/server/server.go [hunk 6]
```diff
@@ -130,13 +164,9 @@ func (m *Server) Run() {
 }
 
 func (m *Server) RunClients() {
```
Comment (kazhuravlev) [inline pkg/server/server.go]: Must receive `context.Context`

File: pkg/server/server.go [hunk 7]
```diff
@@ -130,13 +164,9 @@ func (m *Server) Run() {
...
+
 	for _, peer := range m.BootPeerAddrs {
-		conn, err := net.Dial("tcp", peer)
-		if err != nil {
-			zap.S().Error("failed to connect to peer: ", err)
-			continue
-		}
 
-		go handleClient(conn)
+		go handleClient(peer)
```
Comment (kazhuravlev) [inline pkg/server/server.go]: must receive `context.Context`

File: pkg/p2p/transport.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"context"
+	"net"
+	"time"
+)
+
+// DefaultTransport implements the default dialing strategy
+var DefaultTransport = Transport{
+	DialContext: (&net.Dialer{
+		Timeout:   30 * time.Second,
```
Comment (kazhuravlev) [inline pkg/p2p/transport.go]: hardcode

File: pkg/p2p/transport.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package p2p
+
+import (
+	"context"
+	"net"
+	"time"
+)
+
+// DefaultTransport implements the default dialing strategy
+var DefaultTransport = Transport{
```
[the remark this answers is not in the document]
Comment (montekki) [reply pkg/p2p/transport.go]: yes but it doesn't live in it's own ```transport``` package, but in the ```p2p```

File: pkg/server/server.go (hunk 1 above)
Comment (montekki) [reply pkg/server/server.go]: no, the idea behind this is following: there is a default ```DialContext``` implementation in the transport that just wraps the ```net.Dialer.DialContext``` and tries to connect to a node.
Here we redefine the ```DefaultTransport.DialContext``` behaviour to try to re-adjust the connection number of the node. So, for instance, if the version of the node we are trying to connect to is ```0.5.5``` we begin with trying to connect with version ```0.5.10```, the connection gets terminated, we try ```0.5.9``` and so on. Most likely we will have to connect to a large variety of combinations of Major.Minor.Patch versions. Up until this day only the Major version is constant and for every existing node in the world is ```0```.

**Pull request merged**